url1f1de5fc71_cooijmanstim_hobo_tgz-pJ8-hobo_TestingEnviromentJ8/src/hobo/graphics/GamePanel.java

```diff
@@ -0,0 +1,208 @@
+							                                                        boolean isSelected,
+							                                                        boolean hasFocus) {
+								Mission m = (Mission)o;
+								JLabel c = new JLabel(m.value+" "+m.source+" - "+m.destination);
+								c.setBackground(ps.missionCompleted(m)
+								                  ? new java.awt.Color(0.5f, 0.5f, 1f)
+								                  : new java.awt.Color(1f, 0.5f, 0.5f));
+								c.setOpaque(true);
+								return c;
+							}
+						});
+						c = l;
+						break;
+					}
+					f.add(c);
+				}
+			}
+			f.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
+			f.pack();
+			f.setVisible(true);
+		} else {
+			players.reflect(s);
+			decks.reflect(s);
+			map.reflect(s);
+			missions.reflect(s);
+			hand.reflect(s);
+		}
+	}
+	
+	public PlayerInteraction getUserInterface() {
+		return new PlayerInteraction() {
+			@Override public Decision askDecision(Player p, State s) {
+				return getDecision();
+			}
+
+			@Override public void tellIllegal (Player p, State s, Decision d, String reason) { message("illegal move: "+reason); }
+			@Override public void tellDraw    (Player p, State s) { message("draw game"   ); }
+			@Override public void tellLoss    (Player p, State s) { message(p+", you lose"); }
+			@Override public void tellWin     (Player p, State s) { message(p+", you win" ); }
+			
+			@Override public void observe(Event e) {}
+		};
+	}
+	
+	
+	public void claim(Railway r) {
+		PlayerState ps = state.currentPlayerState();
+		Color selection = hand.selection();
+		CardBag cards = ps.hand.cardsToClaim(r, selection);
+		// if couldn't afford it, just let the move be illegal so the user will be notified
+		if (cards == null)
+			cards = new CardBag();
+		registerDecision(new ClaimRailwayDecision(ps.handle, r, cards));
+	}
+
+	public void drawMissions() {
+		registerDecision(new DrawMissionsDecision(state.currentPlayer()));
+	}
+
+	public void keepMissions(Set<Mission> ms) {
+		registerDecision(new KeepMissionsDecision(state.currentPlayer(), ms));
+	}
+
+	public void drawCard(Color c) {
+		registerDecision(new DrawCardDecision(state.currentPlayer(), c));
+	}
+	
+	
+	public void message(String s) {
+		System.out.println(s);
+		add(new Toast(s), JLayeredPane.POPUP_LAYER);
+	}
+
+	// here be the thread-hackery that is required to make
+	// PlayerInteraction work for a graphical interface.
+	private final Object lastDecisionLock = new Object();
+	private boolean awaitingDecision = false;
+	private Decision lastDecision = null;
+	
+	public void registerDecision(Decision d) {
+		synchronized (lastDecisionLock) {
+			lastDecision = d;
+			lastDecisionLock.notify();
+		}
+	}
+	
+	public Decision getDecision() {
+		awaitingDecision = true;
+		Decision d = null;
+		try {
+			while (d == null) {
+					synchronized (lastDecisionLock) {
+						lastDecisionLock.wait();
+						d = lastDecision;
+						lastDecision = null;
+					}
+			}
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)